helix-core/src/main/java/org/apache/helix/controller/stages/ManagementModeStage.java

```diff
@@ -53,4 +88,87 @@ private void checkInManagementMode(String clusterName, ManagementControllerDataP
+      type = ClusterManagementMode.Type.CLUSTER_PAUSE;
+      if (!instancesFullyFrozen(enabledLiveInstances, liveInstanceMap, allInstanceMessages)) {
+        status = ClusterManagementMode.Status.IN_PROGRESS;
+      }
+    } else {
+      type = ClusterManagementMode.Type.CONTROLLER_PAUSE;
+    }
+
+    return new ClusterManagementMode(type, status);
+  }
+
+  private boolean instancesFullyFrozen(Set<String> enabledLiveInstances,
+      Map<String, LiveInstance> liveInstanceMap,
+      Map<String, Collection<Message>> allInstanceMessages) {
+    // 1. All live instances are frozen
+    // 2. No pending participant status change message.
+    return enabledLiveInstances.stream().noneMatch(
+        instance -> !LiveInstanceStatus.PAUSED.equals(liveInstanceMap.get(instance).getStatus())
+            || hasPendingMessage(
+            allInstanceMessages.getOrDefault(instance, Collections.emptyList()),
+            MessageType.PARTICIPANT_STATUS_CHANGE));
+  }
+
+  private boolean hasPendingMessage(Collection<Message> messages, MessageType type) {
+    return messages.stream().anyMatch(message -> type.name().equals(message.getMsgType()));
+  }
+
+  private void recordManagementModeStatus(ClusterManagementMode mode, HelixDataAccessor accessor) {
+    // update cluster status
+    PropertyKey statusPropertyKey = accessor.keyBuilder().clusterStatus();
+    ClusterStatus clusterStatus = accessor.getProperty(statusPropertyKey);
+    if (clusterStatus == null) {
+      clusterStatus = new ClusterStatus();
+    }
+    if (!mode.getMode().equals(clusterStatus.getManagementMode())
+        || !mode.getStatus().equals(clusterStatus.getManagementModeStatus())) {
+      // status is different, need to write to metastore
+      clusterStatus.setManagementMode(mode.getMode());
+      clusterStatus.setManagementModeStatus(mode.getStatus());
+      if (!accessor.updateProperty(statusPropertyKey, clusterStatus)) {
+        LOG.error("Failed to update cluster status {}", clusterStatus);
+      }
+    }
+  }
+
+  private void recordManagementModeHistory(ClusterManagementMode mode, PauseSignal pauseSignal,
+      String controllerName, HelixDataAccessor accessor) {
+    // Only record completed status
+    if (!ClusterManagementMode.Status.COMPLETED.equals(mode.getStatus())) {
+      return;
+    }
+
+    // Record a management mode history in controller history
+    String path = accessor.keyBuilder().controllerLeaderHistory().getPath();
+    if (!accessor.getBaseDataAccessor().update(path, oldRecord -> {
+      if (oldRecord == null) {
+        oldRecord = new ZNRecord(PropertyType.HISTORY.toString());
+      }
+      String fromHost = (pauseSignal == null ? null : pauseSignal.getFromHost());
+      String reason = (pauseSignal == null ? null : pauseSignal.getReason());
+      return new ControllerHistory(oldRecord).updateManagementModeHistory(controllerName, mode,
+          fromHost, Instant.now().toEpochMilli(), reason);
```

Comment:
nit, Instant.now().toEpochMilli() should be fetched before the update for a more accurate status change time and avoid repeatedly system calls, since the update method might be retried.